src/test/codegen/enum-bounds-check-derived-idx.rs

```diff
@@ -12,13 +12,19 @@ pub enum Bar {
 // CHECK-LABEL: @lookup_inc
 #[no_mangle]
 pub fn lookup_inc(buf: &[u8; 5], f: Bar) -> u8 {
-    // CHECK-NOT: panic_bounds_check
+    // Canary: we should be able to optimize out the bounds check, but we need
+    // to track the range of the discriminant result in order to be able to do that.
+    // oli-obk tried to add that, but that caused miscompilations all over the place.
+    // CHECK: panic_bounds_check
     buf[f as usize + 1]
```

Comment:
I think what we really need here is range attributes on parameters (which LLVM doesn't currently support).

Presumably, this PR generates IR like:
```ll
define @lookup_inc(..., i8 %f) {
  %0 = alloca i8
  store i8 %f, ptr %0
  %1 = load i8, ptr %0, !range !0
  ...
}

!0 = !{i8 0, i8 4}
```

But that will get optimized early on to something like:

```ll
define @lookup_inc(..., i8 %f) {
  %1 = %f
  ...
}
```
...dropping the range metadata on the floor.

One way to avoid this is to add assumes, like we did before. LLVM itself could also, in theory, add assumes in mem2reg to retain knowledge from allocas it drops.

But it seems like the real underlying problem is the lack of range attributes on parameters. If we could just generate IR like:

```ll
define @lookup_inc(..., i8 range(!0) %f) {
  ...
}

!0 = !{i8 0, i8 4}
```

...then we wouldn't need assumes or fragile `!range` metadata on alloca loads in the body of the function.

I will look into this, although I'm slightly worried that the reason `range` attributes aren't supported on parameters is because parameter attributes currently only accept 0 or 1 integer arguments, and it would take extensive refactoring to make them accept multiple arguments or reference a metadata node. Let's hope that's not the case.